### PR TITLE
New module for Azure eventhub namespace

### PR DESCRIPTION
**Pull request by mmyyrroonn**
##### SUMMARY

- **azure_rm_eventhubnamespace**: create, update and delete an Azure eventhub namespace.
- **azure_rm_eventhubnamespace_facts**: gather facts for an a specific eventhub namespace or all eventhub namespaces. Include showing sas policies.

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

- azure_rm_eventhubnamespace

##### ADDITIONAL INFORMATION

```paste below
2.9
```

### CODE REVIEW

**Comment by MyronFanQiu** [comment]
@zikalino @yuwzho @yungezz Would you please review the PR when you are free? Thank you very much!

File: lib/ansible/modules/cloud/azure/azure_rm_eventhubnamespace.py [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+                type='int'
+            ),
+            kafka_enabled=dict(
+                type='bool'
+            )
+        )
+
+        self.results = dict(
+            changed=False,
+            id=dict()
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_eventhubnamespace.py]: id is str

File: lib/ansible/modules/cloud/azure/azure_rm_eventhubnamespace.py [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+        self.maximum_throughput_units = None
+        self.kafka_enabled = None
+        self.tags = None
+
+        super(AzureRMEventHubNamespace, self).__init__(self.module_arg_spec, supports_tags=True, supports_check_mode=True)
+
+    def exec_module(self, **kwargs):
+        """Main module execution method"""
+
+        for key in self.module_arg_spec.keys():
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_eventhubnamespace.py]: +tags

File: lib/ansible/modules/cloud/azure/azure_rm_eventhubnamespace.py [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+            poller = self.eventhub_client.namespaces.create_or_update(resource_group_name=self.resource_group,
+                                                                      namespace_name=self.name,
+                                                                      parameters=eventhubnamespace)
+            return self.get_poller_result(poller)
+        except self.eventhub_models.ErrorResponseException as exc:
+            self.fail('Error creating or updating Event Hub Namespace {0}: {1}'.format(self.name, str(exc.inner_exception) or str(exc.message) or str(exc)))
+
+    def check_status(self, changed, eventhubnamespace):
+        # Compare sku
+        if self.sku and self.sku != eventhubnamespace.sku.name:
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_eventhubnamespace.py:262]: default value of sku is error-prone. suppose first time user set sku as premium or something, second time not setting sku, so default is basic, premium!=basic, but actually user don't want to change this field.

Comment (mmyyrroonn) [reply lib/ansible/modules/cloud/azure/azure_rm_eventhubnamespace.py:262]: I checked the code. During each update process, if the sku is not set, the current sku will be wrote back explicitly. I deleted the confusing part of the documentation. Thanks!

**Comment by MyronFanQiu** [comment]
@zikalino Hi zim~ Could you help me review the PR when you have time? Thank you very much!

**Comment by MyronFanQiu** [comment]
@yuwzho  Hi yuwei~ Could you help me review the PR? Thank you very much! 😄

**Review by yungezz — APPROVED** [review]

**Comment by MyronFanQiu** [comment]
ready_for_review

File: lib/ansible/modules/cloud/azure/azure_rm_eventhubnamespace_facts.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+        if self.name:
+            response = self.get_item()
+        elif self.resource_group:
+            response = self.list_by_resource_group()
+        else:
+            response = self.list_all()
+        self.results['eventhubnamespaces'] = [self.to_dict(x) for x in response if self.has_tags(x.tags, self.tags)]
+
+        if self.show_sas_policies:
+            self.results['eventhubnamespaces'] = [self.get_sas_policies(x) for x in self.results['eventhubnamespaces']]
```
Comment (yuwzho) [inline lib/ansible/modules/cloud/azure/azure_rm_eventhubnamespace_facts.py]: if self.show_sas_policies set, we add an option ['sas_policies'] in every item in ['eventhubnamespaces'], not just return the list of sas policies

**Review by yuwzho — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/azure/azure_rm_eventhubnamespace_facts.py (hunk 4 above)
Comment (mmyyrroonn) [reply lib/ansible/modules/cloud/azure/azure_rm_eventhubnamespace_facts.py]: Yes, for each item ``` x in ['eventhubnamespaces']```, an ```sas_keys : {all keys}``` pair will be added. Am I misunderstanding the meaning?

File: test/integration/targets/azure_rm_eventhub/tasks/main.yml [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+       - output.eventhubnamespaces[0]['kafka_enabled'] != None
+       - output.eventhubnamespaces[0]['sas_keys'] != None
+       - output.eventhubnamespaces[1]['id'] != None
+       - output.eventhubnamespaces[1]['name'] != None
+       - output.eventhubnamespaces[1]['location'] != None
+       - output.eventhubnamespaces[1]['sku'] != None
+       - output.eventhubnamespaces[1]['is_auto_inflate_enabled'] != None
+       - output.eventhubnamespaces[1]['maximum_throughput_units'] != None
+       - output.eventhubnamespaces[1]['kafka_enabled'] != None
+       - output.eventhubnamespaces[1]['sas_keys'] != None
```
Comment (mmyyrroonn) [inline test/integration/targets/azure_rm_eventhub/tasks/main.yml]: @yuwzho  here~ I may change the key name to sas_policies

File: lib/ansible/modules/cloud/azure/azure_rm_eventhubnamespace_facts.py (hunk 4 above)
Comment (yuwzho) [reply lib/ansible/modules/cloud/azure/azure_rm_eventhubnamespace_facts.py]: this operation will construct a new array and assign to variable

Comment (mmyyrroonn) [reply lib/ansible/modules/cloud/azure/azure_rm_eventhubnamespace_facts.py]: adjust the name and the structure of the code

**Review by yuwzho — APPROVED** [review]

**Comment by MyronFanQiu** [comment]
@mattclay @nitzmahone Hello~ Could you please take some time to review this PR? Thank you very much!

**Review by nitzmahone — CHANGES_REQUESTED** [review]
Looks fine to me, other than `_facts` module suffix needs to be changed to `_info` before merging for 2.9. Also wondering if this one might be a good candidate for trying Zim's new lightweight module pattern...

**Comment by zikalino** [comment]
@nitzmahone yes, I agree we should start merging _info modules.
regarding lightweight pattern, my idea is that ppl should continue in their own way right now,
so i think we should merge this one as is.
and we should start with generated modules.
I also plan to write detailed guidelines that everybody has to agree to.

**Comment by mattclay** [comment]
bot_status

**Comment by Fred-sun** [comment]
@MyronFanQiu Please complete the PR and promote the merger when you are free, thank you very much!

**Comment by Fred-sun** [comment]
@MyronFanQiu

**Comment by MyronFanQiu** [comment]
@Fred-sun Hello. Thanks for your help! we should add `shippable/azure/group1` or `shippable/azure/group2` instead of adding `unsupported`. We still want to have a test for this module.

**Comment by Fred-sun** [comment]
@MyronFanQiu Ok, I will update this.  What are the differences between these groups(shippable/azure/group(1|2|3|4|5|6|7|8|9|10)? Thanks a lot!

**Comment by Fred-sun** [comment]
@MyronFanQiu There is a test timeout, please help to trigger CI check again! Thanks a lot!

**Comment by Fred-sun** [comment]
ready_for_review

**Comment by Fred-sun** [comment]
@yungezz @zikalino There are other PR（#55509 related to this PR , please review and promote the merger. Thank you very much!

**Comment by Fred-sun** [comment]
@MyronFanQiu @yungezz This PR already approved. Now this PR show exist conflicts, but I don't see the cause of the conflict. Could you take a look this and promote to merge? Thank you very much!

**Comment by Fred-sun** [comment]
@MyronFanQiu Can you find out the reason for CI fail? I don't know how to fix this CI fail, thank you!

**Comment by Fred-sun** [comment]
@MyronFanQiu Please re-trigger CI check! Thanks a lot!

**Comment by Fred-sun** [comment]
@MyronFanQiu Could you help resolve the CI fail? I tried, but I couldn't solve it. Thank you very much!

**Comment by Fred-sun** [comment]
@MyronFanQiu Could you help fixes the CI fail when you're free! Thank you very much!

**Pull request closed without merging**